api-resources.md

```diff
@@ -0,0 +1,414 @@
+  HpkePublicKey public_key;
+} HpkeConfig;
+```
+
+#### Required HTTP methods
+
+##### GET/HEAD
+
+The response body is a `struct HpkeConfig` representing the `hpke_config`.
+
+### Reports
+
+#### Path
+
+`/tasks/task-id/reports/report-id`
+
+`report-id` uniquely identifies a report in a task. It is chosen by the client when uploading a report.
+
+#### Representation
+
+```
+struct {
+  Time time;
+  Extension extensions<0..2^16-1>;
+} ReportMetadata;
+
+struct {
+  ReportMetadata metadata;
+  opaque public_share<0..2^32-1>;
+  HpkeCiphertext encrypted_input_shares<1..2^32-1>;
+} Report;
+```
+
+#### Required HTTP methods
+
+##### GET/HEAD
+
+The response body is a `struct Report` representing the report.
+
+##### PUT
+
+Idempotent upload of a report to an aggregator. The request body is a `struct Report`.
+
+This is analogous to DAP-02's `POST /upload`.
+
+### An aggregation job
+
+Only the helper supports this resource.
+
+#### Path
+
+`/tasks/task-id/aggregation_jobs/aggregation-job-id`
+
+`aggregation-job-id` uniquely identifies an aggregation job. It is chosen by the leader when it dispatches a job to the helper.
+
+#### Representation
+
+```
+enum {
+  continued(0),
+  finished(1),
+  failed(2),
+  (255),
+} PrepareStepState;
+
+struct {
+  ReportID report_id;
+  PrepareStepState prepare_step_state;
+  select (PrepareStep.prepare_step_state) {
+    case continued: opaque prep_msg<0..2^32-1>; /* VDAF preparation message */
+    case finished: Empty;
+    case failed: ReportShareError;
+  }
+} PrepareStep;
+
+struct {
+  u16 round;
+  PrepareStep prepare_steps<1..2^32-1>;
+} AggregateJob
+```
+
+See "POST", below, for explanation and justification of the `round` field.
+
+#### Required HTTP methods
+
+##### GET/HEAD
+
+The response body is a `struct AggregateJob` representing the current state of the reports in the job.
+
+##### PUT
+
+PUT to an aggregate job is how the leader creates an aggregate job in the helper. The body is an `AggregateJobPutReq`:
+
+```
+struct {
+  ReportMetadata metadata;
+  opaque public_share<0..2^32-1>;
+  HpkeCiphertext encrypted_input_share;
+} ReportShare;
+
+struct {
+  QueryType query_type;
+  select (PartialBatchSelector.query_type) {
+    case time_interval: Empty;
+    case fixed_size: BatchID batch_id;
+  };
+} PartialBatchSelector;
+
+struct {
+  opaque agg_param<0..2^16-1>;
+  PartialBatchSelector part_batch_selector;
+  ReportShare report_shares<1..2^32-1>;
+} AggregateJobPutReq;
+```
+
+If successful, the helper's response is a `struct AggregateJob` where `prepare_steps` contains the helper's first-round VDAF prepare messages for the report shares in `AggregateJobPutReq`, or errors if `vdaf.prepare` failed for any report.
+
+This is analogous to DAP-02's `POST /aggregate` with an `AggregateInitReq`.
+
+This request is idempotent, so if a helper receives multiple PUT requests for some aggregation job and the members of `AggregateJobPutReq` don't change, the helper can return 200 OK.
```

Comment:
I don't think this is idempotent if aggregator proactively aggregates reports. If two identical AggregateJobPutReq is received before and after a batch is collected, then the later one will end up in error.

Reply:
even without collect flow, anti-reply will make this non-idempotent too:

https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/blob/main/draft-ietf-ppm-dap.md#early-input-share-validation-early-input-share-validation

Reply:
It's idempotent in the sense that repeated `PUT /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` requests won't change the state, the key point being that it's safe for the leader to resend that request to the helper multiple times.

I also think that collection of a batch doesn't need to affect how the helper responds to this request. Let me sketch out a hypothetical:

1. Leader sends `PUT /tasks/{task-id}/aggregation_jobs/agg-job-1` (here, `agg-job-1` is some specific aggregation job identifier).
2. Helper replies with `struct AggregateJob` containing first-round prepare message
3. Leader sends `POST /tasks/{task-id}/aggregation_jobs/agg-job-1` with first-round broadcast prepare message
4. Helper prepares the reports, consuming batch query count for the relevant reports, and replies with `struct AggregateJob` containing a sequence of `PrepareStep` that are all in state `finished`.

At this point, if the leader sends `PUT /tasks/{task-id}/aggregation_jobs/agg-job-1` again (for whatever reason) _and if all the members of the `struct AggregateJobPutReq` are unchanged_, then the helper shouldn't re-evaluate anti-replay, because it can check its own storage, notice it's already prepared all those inputs, and return the same response it did in step (4). This doesn't violate anti-replay or privacy budgets because it's the exact same set of reports under the same aggregation parameter.

However, if the leader sends `PUT /tasks/{task-id}/aggregation_jobs/agg-job-1` _but varies any of the fields of the body_, then the helper should reject it. So I think we should specify that the aggregation job resource may not be mutated once created, except to advance its state through a POST request.

Further, if the leader sends `PUT /tasks/{task-id}/aggregation_jobs/agg-job-2` and the reports in that request's body overlap with the ones prepared by `agg-job-1`, then the helper has to apply anti-replay and either consume query limit or refuse to service the request, as appropriate.

Reply:
@tgeoghegan makes sense, I think it's safe privacy wise since the agg flow is for preparation. My other concern is time related states, if the 2nd `PUT /tasks/{task-id}/aggregation_jobs/agg-job-1` come at a time the task_id has expired, would helper responds differently? If not then we are potentially asking helper to store output share or aggregate share forever.

Reply:
From DAP-02 [5.4.1](https://www.ietf.org/archive/id/draft-ietf-ppm-dap-02.html#section-5.4.1):

> Furthermore, the aggregators must store data related to a task as long as the current time has not passed this task's task_expiration. Aggregator MAY delete the task and all data pertaining to this task after task_expiration. Implementors SHOULD provide for some leeway so the collector can collect the batch after some delay.

So if a task expires, then from that point, a helper MAY respond to `PUT /tasks/{task-id}/aggregation_jobs/agg-job-1` with an error. It could also choose to continue providing the results of the aggregation job indefinitely. It's up to the implementation to decide when it discards data belonging to a task like aggregate results.

I think there will have to be some text explaining that leaders should be prepared for helpers to have discarded the results of an aggregation job, but I don't think this is an idempotence problem, because the state change (helper discarding aggregate job results) wasn't a result of repeated `PUT` requests by the leader.

Reply:
I misunderstood idempotency, your explanation is correct: it's not about the response being the same but the state not changed by multiple calls.